### ISSUE TITLE
Why does docker run cause portlistening on 0.0.0.0 instead of 127.0.0.1 after Upgrading to 1.6.2?

### ISSUE THREAD

**Issue by PretoPlasma** [rancher-sandbox/rancher-desktop#3389]
### Actual Behavior

Last week I updated my rancher-desktop from 1.5.1 to 1.6.2.

Since then my containers running can not be reached via 127.0.0.1.

I investigated this issue and found out that the ports defined in the docker run command are mapped to 0.0.0.0 instead of 127.0.0.1 during container initialization. A docker container logs --follow xxx shows, that the container is running well.

Does anyone else face this issue?

I already  uninstalled rancher-desktop and removed the wsl distributions (wsl --unregister) then installed rancher-destkop V1.5.1 from scratch. However the given ports are stilled bound to 0.0.0.0 which makes my container unavailable locally.

Does anyone understand my problem? Can anyone help?

Any tipp or help is highly appreciated.

### Steps to Reproduce

After starting the container with e.c. (portainer):

docker run -d -p 8000:8000 -p 9443:9443 --name portainer --restart=always -v /var/run/docker.sock:/var/run/docker.sock -v portainer_data:/data portainer/portainer-ce:latest

netstat -anop TCP lists

TCP    0.0.0.0:9443           0.0.0.0:0              ABHÖREN         xxx
  
instead of
  
TCP    127.0.0.1:9443         0.0.0.0:0              ABHÖREN         xxx

which should be the correct port to ip mapping.

a docker container logs --follow portainer demonstrates, that portainer is running well.

This is just an example. My other containers react in the same manner: netstats displays portmapping on 0.0.0.0 instead of 127.0.0.1 and container logs shows a well running container.

### Result

The result is, that e.c. the portainer website https://localhost:9443 is not reachable.
The browser displays the error message (translated from german to english):

"The website is temporarily not available or was permanently moved to a new web adress.
ERR_CONNECTION_ABORTED."

The container logs show a well running container.

### Expected Behavior

The container website e.c. portainer https://localhost:9443 displays correctly without the unavailable screen.

### Additional Information

I am not sure if this is related to the update to V1.6.2. But this was the last action I was performing before the error occured.

### Rancher Desktop Version

1.6.2

### Rancher Desktop K8s Version

Kubernetes is disabled (1.23.6)

### Which container engine are you using?

moby (docker cli)

### What operating system are you using?

Windows

### Operating System / Build Version

Windows 10 Pro Version 21H2 Build 19044.1708

### What CPU architecture are you using?

x64

### Windows User Only

I am using VPN Cisco Any Connect, Trend Micro Apex One

**Comment by Nino-K** [comment]
@PretoPlasma RD desktop listening on `127.0.0.1` by default was a bug and not a feature as reported by our users here: https://github.com/rancher-sandbox/rancher-desktop/issues/1180.
However, this bug was addressed in our releases from `1.6.0` and onward. 

I went ahead and tested `portainer` as you mentioned above over cisco anyconnect (although VPN is not playing any role in this issue) and everything seemed to be working. When a port is bound to `0.0.0.0` it means that it can be accessed via all the IP addresses on all interfaces on the machine, including `127.0.0.1`. If you would like to access your application explicitly on `127.0.0.1` you can always define that address on your docker run command e.g `docker run ... -p 127.0.0.1:9443`.

I suspect the issue here might not be related to the rancher desktop but rather a SSL certificate issue, I would recommend either trying to reset your browser to access the application or even trying a different browser to see if that is the case.

**Comment by PretoPlasma** [comment]
@Nino-K Thanks for your quick reply!

I compared my environment and the netstat output with the one from my colleagues and the only difference I found was that RD listens on 127.0.0.1 instead of 0.0.0.0. So I assumed that this could be a cause but you are right: if 0.0.0.0 stands for ALL ip-adresses it should listen on 127.0.0.1 too.

Your advise to precise port matching on docker run command did not work for me. According to https://docs.docker.com/config/containers/container-networking/ the given ip-adress defines the host ip not from where to listen.

Your advise to reset the browser did not work for me. The container websites are still not available via 127.0.0.1:port, although the browser is clean and fresh. A curl -sv http://localhost:port on wsl console returns the website correctly. So I assume there must be something going wron in between the NAT from windows to wsl.

I allready contacted my it-department to figure it out, however they would argue first to check that this is not related to RD desktop. So from this point of investigation can we be 100% sure that the installation / update of RD Desktop does not have anything to do with this issue?

**Comment by PretoPlasma** [comment]
In addition to my last comment I found out that for some reason svchost.exe keeps the assigned container ports alive, after stopping RD Desktop. A restart did not prevent this. After restarting windows this svchost.exe resurrects magically keeping the assigned container ports alive. Only a pskill can free up these ports. However after restarting RD Desktop another svchost.exe keeps these ports alive in the same manner. Can you please check this for me?

**Comment by Nino-K** [comment]
> In addition to my last comment I found out that for some reason svchost.exe keeps the assigned container ports alive, after shutting down RD Desktop. A restart of windows did not stop this behaviour.
> 
> `netstat -anop TCP | findstr 9443 TCP 0.0.0.0:9443 0.0.0.0:0 LISTEN 17960'
> 
> After restarting windows another svchost.exe is raised magically which then keeps the assigned container ports alive again! Only a pskill could free up these ports.
> 
> `pskill 17960 netstat -anop TCP | findstr 9443

I have managed to reproduce this myself and created an issue for it here: https://github.com/rancher-sandbox/rancher-desktop/issues/3399

However, I would like to emphasize that this is a separate issue and it's independent of your original issue.

**Comment by PretoPlasma** [comment]
@Nino-K I agree with you. These are two separate issues. I setup the rancher desktop environment from scratch however it still seem that the ports are not mapped to the containers correctly. Running my container's webapp directly on nodejs the site is opened as aspected. Running the webapp in the container, within wsl it is accessible, outside wsl - so from my pc's browser - it is not accessible.

**Comment by PretoPlasma** [comment]
@Nino-K I found out that the portmapping which is boot resistant is envoked and kept alive by the iphlpsvc. This can be found by netstat -anop TCP |findstr port. The PID on my system is every time the same 3840 which is the iphlpsvc-service. Killing the service leads in a restart by the operating system. My question is: how did this service get notice about my container ports. I did not find anything in the registry HKEY_LOCAL_MACHINE\ SYSTEM\ CurrentControlSet\ Services\ iphlpsvc. So there must be another storage for this information. And there must be communication between wsl / rancher desktop and this service. Do you agree?

**Comment by Nino-K** [comment]
@PretoPlasma yes, the fix will be available in our next release, which should be coming up soon.

**Comment by Nino-K** [comment]
@PretoPlasma I went ahead and tested the original issue you mentioned one more time and below are the steps I did:

```
docker run -d -p 8000:8000 -p 9443:9443 --name portainer --restart=always -v /var/run/docker.sock:/var/run/docker.sock -v portainer_data:/data portainer/portainer-ce:latest
```
Looking at the netsh:
```
PS C:\Users\XYZ> netsh interface portproxy show all

Listen on ipv4:             Connect to ipv4:

Address         Port        Address         Port
--------------- ----------  --------------- ----------
0.0.0.0         8000        172.17.237.247  8000
0.0.0.0         9443        172.17.237.247  9443

Listen on ipv6:             Connect to ipv6:

Address         Port        Address         Port
--------------- ----------  --------------- ----------
::              8000        fe80::215:5dff:feff:9f56 8000
::              9443        fe80::215:5dff:feff:9f56 9443
```
I can access the portainer on both localhost and `127.0.0.1` (please look at the attachments).

Inside the WSL VM:
`wsl -d rancher-desktop`
```
/mnt/c/Users/XYZ # netstat -nlp | grep 9443
tcp        0      0 0.0.0.0:9443            0.0.0.0:*               LISTEN      31419/docker-proxy
tcp        0      0 :::9443                 :::*                    LISTEN      31427/docker-proxy
```
Now curling our endpoint from inside WSL VM:
```
/mnt/c/Users/XYZ # curl -I -k https://127.0.0.1:9443
HTTP/2 307
content-type: text/html; charset=utf-8
location: /timeout.html
date: Tue, 29 Nov 2022 23:58:19 GMT
```
**I also test it by specifying `127.0.0.1`:**

```
docker run -d -p 8000:8000 -p 127.0.0.1:9443:9443 --name portainer --restart=always -v /var/run/docker.sock:/var/run/docker.sock -v portainer_data:/data portainer/portainer-ce:latest
```
Looking at the netsh on the host:

```
PS C:\Users\XYZ> netsh interface portproxy show all

Listen on ipv4:             Connect to ipv4:

Address         Port        Address         Port
--------------- ----------  --------------- ----------
0.0.0.0         8000        172.17.237.247  8000
127.0.0.1       9443        172.17.237.247  9443

Listen on ipv6:             Connect to ipv6:

Address         Port        Address         Port
--------------- ----------  --------------- ----------
::              8000        fe80::215:5dff:feff:9f56 8000
```
Notice this time netsh is showing `127.0.0.1`.

Inside the WSL VM:
`wsl -d rancher-desktop`
```
/mnt/c/Users/XYZ # netstat -nlp | grep 9443
tcp        0      0 127.0.0.1:9443          0.0.0.0:*               LISTEN      5151/docker-proxy
```
Now curling our endpoint from inside wsl VM:
```
/mnt/c/Users/XYZ # curl -I -k https://127.0.0.1:9443
HTTP/2 200
accept-ranges: bytes
cache-control: max-age=31536000
content-type: text/html; charset=utf-8
last-modified: Mon, 21 Nov 2022 00:37:50 GMT
vary: Accept-Encoding
x-content-type-options: nosniff
x-xss-protection: 1; mode=block
content-length: 18201
date: Wed, 30 Nov 2022 00:03:19 GMT
```
![Screen Shot 2022-11-29 at 3 56 56 PM](https://user-images.githubusercontent.com/10409174/204675578-9313cff0-94a1-4f61-9353-d268c4262014.png)


Everything works as expected, please let me know if I'm missing something here.

**Comment by ry-suzuki** [comment]
I have a similar problem. Since upgrading to 1.6.2, it does not always happen, but access via 127.0.0.1 sometimes fails.
Note that firewall is blocking outgoing connections in my environment.
```
netsh advfirewall firewall add rule name= "Block Outbound Port 8080" dir=out action=block protocol=TCP remoteport=8080
```

If access fails, 127.0.0.1:8080 is not explicitly opened.
```
PS C:\> docker run --rm -d -p 8080:80 nginx
027099d80bac2facacce60645bf5a5d4ea7e3e9ac19f8bcaeb9d94b09d4ce552

PS C:\> curl.exe -I http://127.0.0.1:8080
curl: (56) Recv failure: Connection was reset

PS C:\> netstat -aon | findstr 8080
  TCP         0.0.0.0:8080           0.0.0.0:0              LISTENING       3960
  TCP         [::]:8080              [::]:0                 LISTENING       3960
  TCP         [::1]:8080             [::]:0                 LISTENING       25940

PS C:\> tasklist /svc | findstr "3960 25940"
svchost.exe                   3960 iphlpsvc
wslhost.exe                  25940 N/A
```

Connection to the WSL VM is blocked by the firewall.
```
PS C:\> netsh interface portproxy show all | findstr 8080
0.0.0.0         8080        172.28.123.222  8080
::              8080        fe80::215:5dff:fe25:4e47 8080

PS C:\> curl.exe -I http://172.28.123.222:8080
curl: (7) Failed to connect to 172.28.123.222 port 8080 after 0 ms: Bad access
```

Inside the WSL VM, access is successful.
```
PS C:\> wsl -d rancher-desktop netstat -lnp | findstr 8080
tcp        0      0 0.0.0.0:8080            0.0.0.0:*               LISTEN      4870/docker-proxy
tcp        0      0 :::8080                 :::*                    LISTEN      4878/docker-proxy

PS C:\> wsl -d rancher-desktop curl -I http://127.0.0.1:8080
HTTP/1.1 200 OK
Server: nginx/1.23.2
Date: Tue, 20 Dec 2022 09:08:58 GMT
Content-Type: text/html
Content-Length: 615
Last-Modified: Wed, 19 Oct 2022 07:56:21 GMT
Connection: keep-alive
ETag: "634fada5-267"
Accept-Ranges: bytes

PS C:\> wsl -d rancher-desktop curl -I http://172.28.123.222:8080
HTTP/1.1 200 OK
Server: nginx/1.23.2
Date: Tue, 20 Dec 2022 09:09:01 GMT
Content-Type: text/html
Content-Length: 615
Last-Modified: Wed, 19 Oct 2022 07:56:21 GMT
Connection: keep-alive
ETag: "634fada5-267"
Accept-Ranges: bytes
```

On the other hand, if the access succeeds, 127.0.0.1:8080 is opened by wslhost.exe.
```
PS C:\> docker run --rm -d -p 8080:80 nginx
f7d99d82e892534df984edaf88911540f031ef14893e78b5013b381603a9e83f

PS C:\> curl.exe -I http://127.0.0.1:8080
HTTP/1.1 200 OK
Server: nginx/1.23.2
Date: Tue, 20 Dec 2022 09:37:48 GMT
Content-Type: text/html
Content-Length: 615
Last-Modified: Wed, 19 Oct 2022 07:56:21 GMT
Connection: keep-alive
ETag: "634fada5-267"
Accept-Ranges: bytes

PS C:\> netstat -aon | findstr 8080
  TCP         0.0.0.0:8080           0.0.0.0:0              LISTENING       3960
  TCP         127.0.0.1:8080         0.0.0.0:0              LISTENING       25940
  TCP         127.0.0.1:55218        127.0.0.1:8080         TIME_WAIT       0
  TCP         [::]:8080              [::]:0                 LISTENING       3960
  TCP         [::1]:8080             [::]:0                 LISTENING       25940

PS C:\> tasklist /svc | findstr "3960 25940"
svchost.exe                   3960 iphlpsvc
wslhost.exe                  25940 N/A
```

Connection to the WSL VM is still blocked.
```
PS C:\> netsh interface portproxy show all | findstr 8080
0.0.0.0         8080        172.28.123.222  8080
::              8080        fe80::215:5dff:fe25:4e47 8080

PS C:\> curl.exe -I http://172.28.123.222:8080
curl: (7) Failed to connect to 172.28.123.222 port 8080 after 0 ms: Bad access
```


As a side note, even if firewall is not blocking, access via 127.0.0.1 can fail if the interface to bind to is specified as 127.0.0.1 only.
```
PS C:\> netsh advfirewall firewall delete rule name= "Block Outbound Port 8080"

1 規則を削除しました。
OK
```
```
PS C:\> docker run --rm -d -p 127.0.0.1:8080:80 nginx
aae98a3a8ab71c70cca3884a3dc996075e05a1d310ab046e7ce771b763bed63a

PS C:\> curl.exe -I http://127.0.0.1:8080
curl: (56) Recv failure: Connection was reset
```

127.0.0.1:8080 is opened by iphlpsvc.
```
PS C:\> netstat -aon | findstr 8080
  TCP         127.0.0.1:8080         0.0.0.0:0              LISTENING       3960

PS C:\> tasklist /svc | findstr "3960"
svchost.exe                   3960 iphlpsvc

PS C:\> netsh interface portproxy show all | findstr 8080
127.0.0.1       8080        172.28.123.222  8080
```

Connection to the WSL VM is not blocked by the firewall, but the WSL VM only opens 127.0.0.1:8080, so connection is refused.
```
PS C:\> curl.exe -I http://172.28.123.222:8080
curl: (7) Failed to connect to 172.28.123.222 port 8080 after 2051 ms: Connection refused

PS C:\> wsl -d rancher-desktop netstat -lnp | findstr 8080
tcp        0      0 127.0.0.1:8080          0.0.0.0:*               LISTEN      8790/docker-proxy

PS C:\> wsl -d rancher-desktop curl -I http://127.0.0.1:8080
HTTP/1.1 200 OK
Server: nginx/1.23.2
Date: Tue, 20 Dec 2022 10:29:32 GMT
Content-Type: text/html
Content-Length: 615
Last-Modified: Wed, 19 Oct 2022 07:56:21 GMT
Connection: keep-alive
ETag: "634fada5-267"
Accept-Ranges: bytes

PS C:\> wsl -d rancher-desktop curl -I http://172.28.123.222:8080
curl: (7) Failed to connect to 172.28.123.222 port 8080 after 0 ms: Connection refused
```

**Comment by ry-suzuki** [comment]
I have a similar problem. Since upgrading to 1.6.2, it does not always happen, but access via 127.0.0.1 sometimes fails.
I think the behavior depends on whether the WSL builtin forwarding works or not.

Note that firewall is blocking outgoing connections in my environment.
```
netsh advfirewall firewall add rule name= "Block Outbound Port 8080" dir=out action=block protocol=TCP remoteport=8080
```

If access fails, 127.0.0.1:8080 is not explicitly opened.
```
PS C:\> docker run --rm -d -p 8080:80 nginx
027099d80bac2facacce60645bf5a5d4ea7e3e9ac19f8bcaeb9d94b09d4ce552

PS C:\> curl.exe -I http://127.0.0.1:8080
curl: (56) Recv failure: Connection was reset

PS C:\> netstat -aon | findstr 8080
  TCP         0.0.0.0:8080           0.0.0.0:0              LISTENING       3960
  TCP         [::]:8080              [::]:0                 LISTENING       3960
  TCP         [::1]:8080             [::]:0                 LISTENING       25940

PS C:\> tasklist /svc | findstr "3960 25940"
svchost.exe                   3960 iphlpsvc
wslhost.exe                  25940 N/A
```

Connection to the WSL VM is blocked by the firewall.
```
PS C:\> netsh interface portproxy show all | findstr 8080
0.0.0.0         8080        172.28.123.222  8080
::              8080        fe80::215:5dff:fe25:4e47 8080

PS C:\> curl.exe -I http://172.28.123.222:8080
curl: (7) Failed to connect to 172.28.123.222 port 8080 after 0 ms: Bad access
```

Inside the WSL VM, access is successful.
```
PS C:\> wsl -d rancher-desktop netstat -lnp | findstr 8080
tcp        0      0 0.0.0.0:8080            0.0.0.0:*               LISTEN      4870/docker-proxy
tcp        0      0 :::8080                 :::*                    LISTEN      4878/docker-proxy

PS C:\> wsl -d rancher-desktop curl -I http://127.0.0.1:8080
HTTP/1.1 200 OK
Server: nginx/1.23.2
Date: Tue, 20 Dec 2022 09:08:58 GMT
Content-Type: text/html
Content-Length: 615
Last-Modified: Wed, 19 Oct 2022 07:56:21 GMT
Connection: keep-alive
ETag: "634fada5-267"
Accept-Ranges: bytes

PS C:\> wsl -d rancher-desktop curl -I http://172.28.123.222:8080
HTTP/1.1 200 OK
Server: nginx/1.23.2
Date: Tue, 20 Dec 2022 09:09:01 GMT
Content-Type: text/html
Content-Length: 615
Last-Modified: Wed, 19 Oct 2022 07:56:21 GMT
Connection: keep-alive
ETag: "634fada5-267"
Accept-Ranges: bytes
```

On the other hand, if the access succeeds, 127.0.0.1:8080 is opened by wslhost.exe.
```
PS C:\> docker run --rm -d -p 8080:80 nginx
f7d99d82e892534df984edaf88911540f031ef14893e78b5013b381603a9e83f

PS C:\> curl.exe -I http://127.0.0.1:8080
HTTP/1.1 200 OK
Server: nginx/1.23.2
Date: Tue, 20 Dec 2022 09:37:48 GMT
Content-Type: text/html
Content-Length: 615
Last-Modified: Wed, 19 Oct 2022 07:56:21 GMT
Connection: keep-alive
ETag: "634fada5-267"
Accept-Ranges: bytes

PS C:\> netstat -aon | findstr 8080
  TCP         0.0.0.0:8080           0.0.0.0:0              LISTENING       3960
  TCP         127.0.0.1:8080         0.0.0.0:0              LISTENING       25940
  TCP         127.0.0.1:55218        127.0.0.1:8080         TIME_WAIT       0
  TCP         [::]:8080              [::]:0                 LISTENING       3960
  TCP         [::1]:8080             [::]:0                 LISTENING       25940

PS C:\> tasklist /svc | findstr "3960 25940"
svchost.exe                   3960 iphlpsvc
wslhost.exe                  25940 N/A
```

Connection to the WSL VM is still blocked.
```
PS C:\> netsh interface portproxy show all | findstr 8080
0.0.0.0         8080        172.28.123.222  8080
::              8080        fe80::215:5dff:fe25:4e47 8080

PS C:\> curl.exe -I http://172.28.123.222:8080
curl: (7) Failed to connect to 172.28.123.222 port 8080 after 0 ms: Bad access
```


As a side note, even if firewall is not blocking, access via 127.0.0.1 can fail if the interface to bind to is specified as 127.0.0.1 only.
```
PS C:\> netsh advfirewall firewall delete rule name= "Block Outbound Port 8080"

1 規則を削除しました。
OK
```
```
PS C:\> docker run --rm -d -p 127.0.0.1:8080:80 nginx
aae98a3a8ab71c70cca3884a3dc996075e05a1d310ab046e7ce771b763bed63a

PS C:\> curl.exe -I http://127.0.0.1:8080
curl: (56) Recv failure: Connection was reset
```

127.0.0.1:8080 is opened by iphlpsvc.
```
PS C:\> netstat -aon | findstr 8080
  TCP         127.0.0.1:8080         0.0.0.0:0              LISTENING       3960

PS C:\> tasklist /svc | findstr "3960"
svchost.exe                   3960 iphlpsvc

PS C:\> netsh interface portproxy show all | findstr 8080
127.0.0.1       8080        172.28.123.222  8080
```

Connection to the WSL VM is not blocked by the firewall, but the WSL VM only opens 127.0.0.1:8080, so connection is refused.
```
PS C:\> curl.exe -I http://172.28.123.222:8080
curl: (7) Failed to connect to 172.28.123.222 port 8080 after 2051 ms: Connection refused

PS C:\> wsl -d rancher-desktop netstat -lnp | findstr 8080
tcp        0      0 127.0.0.1:8080          0.0.0.0:*               LISTEN      8790/docker-proxy

PS C:\> wsl -d rancher-desktop curl -I http://127.0.0.1:8080
HTTP/1.1 200 OK
Server: nginx/1.23.2
Date: Tue, 20 Dec 2022 10:29:32 GMT
Content-Type: text/html
Content-Length: 615
Last-Modified: Wed, 19 Oct 2022 07:56:21 GMT
Connection: keep-alive
ETag: "634fada5-267"
Accept-Ranges: bytes

PS C:\> wsl -d rancher-desktop curl -I http://172.28.123.222:8080
curl: (7) Failed to connect to 172.28.123.222 port 8080 after 0 ms: Connection refused
```

**Comment by Nino-K** [comment]
This should no longer be an issue in our latest release. I will go ahead and close this issue, you can always re-open if you encounter anything further.
